kmk/common/internal_keycodes.py

```diff
@@ -16,13 +16,54 @@ def process_internal_key_event(state, action, changed_key, logger=None):
         return tg(state, action, changed_key, logger=logger)
     elif changed_key.code == Keycodes.Layers._KC_TO:
         return to(state, action, changed_key, logger=logger)
+    elif changed_key.code == Keycodes.KMK.KC_GESC.code:
+        return grave_escape(action, state, logger=logger)
     else:
         return state
 
 
-def tilde(state, action, changed_key, logger):
-    # TODO Actually process keycodes
-    return state
+def grave_escape(action, state, logger):
+    if action['type'] == KEY_DOWN_EVENT:
+        for key in state.keys_pressed:
+            if (key == Keycodes.Modifiers.KC_LSHIFT or
+                    key == Keycodes.Modifiers.KC_RSHIFT):
+                # if Shift is held, return KC_GRAVE which will become KC_TILDE on OS level
+                newstate = state.update(
+                    keys_pressed=(
+                        state.keys_pressed | {Keycodes.Common.KC_GRAVE}
+                    ),
+                )
+                return newstate
+            elif (key == Keycodes.Modifiers.KC_LGUI or
+                    key == Keycodes.Modifiers.KC_RGUI):
+                # if GUI is held, return KC_GRAVE
+                newstate = state.update(
+                    keys_pressed=(
+                        state.keys_pressed | {Keycodes.Common.KC_GRAVE}
+                    ),
+                )
+                return newstate
+            else:
+                # else return KC_ESC
+                newstate = state.update(
+                    keys_pressed=(
+                        state.keys_pressed | {Keycodes.CodeKeys.CODE_ESC}
+                    ),
+                )
+                return newstate
+
+    elif action['type'] == KEY_UP_EVENT:
+        newstate = state.update(
```

Comment:
Can flatten this into a single `update` call by checking for `if key not in (CODE_ESC, KC_GRAVE)`, but this technically still works. Turns an `O(n^2)` operation into a `2O(n^2)`, though, which may matter on weaker chips (but still... probably not).